javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala

```diff
@@ -0,0 +1,244 @@
+package java.util.concurrent
+
+import java.util._
+import scala.collection.JavaConversions._
+
+trait ConcurrentMap[K, V] extends java.util.Map[K, V] {
+
+  def remove(key: Any, value: Any): Boolean = {
+    if (containsKey(key) && get(key) === value) {
+      remove(key)
+      true
+    } else false
+  }
+
+  def replace(key: K, value: V): V = {
+    if (containsKey(key)) {
+      put(key, value)
+    } else get(key)
+  }
+
+  def replace(key: K, oldValue: V, newValue: V): Boolean = {
+    if (containsKey(key) && get(key) === oldValue) {
+      put(key, newValue)
+      true
+    } else false
+  }
+}
+
+class ConcurrentHashMap[K >: Null, V >: Null](initialCapacity: Int,
+    loadFactor: Float,
+    concurrencyLevel: Int)
+    extends java.util.AbstractMap[K, V]
+    with java.util.concurrent.ConcurrentMap[K, V]
+    with java.io.Serializable { self =>
+
+  def this() = 
+    this(16, 0.75F, 16)
+
+  def this(initialCapacity: Int) = 
+    this(initialCapacity, 0.75F, 16)
+
+  def this(initialCapacity: Int, loadFactor: Float) = 
+    this(initialCapacity, loadFactor, 16)
+
+  def this(initialMap: java.util.Map[_ <: K, _ <: V]) = {
+    this(16, 0.75F, 16)
+    putAll(initialMap)
+  }
+
+  private final case class Box[K](inner: K) {
+
+    def apply(): K = inner
+
+    override def equals(o: Any): Boolean = {
+      o match {
+        case b: Box[_] => inner === b.inner
+        case any => any === inner 
+      }
+    }
+  }
+
+  private val inner = new scala.collection.mutable.HashMap[Box[K], V]()
+
+  override def clear(): Unit =
+    inner.clear
+
+  override def containsKey(key: Any): Boolean =
+    inner.exists { case (ik, _) => key === ik() }
+
+  override def containsValue(value: Any): Boolean =
+    inner.exists { case (_, iv) => value === iv }
+    
+  def elements(): java.util.Enumeration[V] =
+    asJavaEnumeration(inner.values.iterator)
+
+  def entrySet(): java.util.Set[java.util.Map.Entry[K, V]] = {
+    new java.util.AbstractSet[java.util.Map.Entry[K, V]] {
+      override def size(): Int = self.size
+
+      val keysCopy: collection.Iterator[Box[K]] = inner.keysIterator
```

Comment:
This should be inside the `def iterator()` method. Otherwise you'll reuse the same underlying iterator for all new iterators, which means all but the first one will be effectively empty, having been exhausted by the first underlying iterator.